### PR TITLE
KAFKA-15776: Introduce remote.fetch.max.timeout.ms to configure DelayedRemoteFetch timeout

### PR DESCRIPTION
**Pull request by kamalcph**
[KIP-1018](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1018%3A+Introduce+max+remote+fetch+timeout+config+for+DelayedRemoteFetch+requests)

Test: Existing unit and integration tests

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 1]
```diff
@@ -1338,7 +1338,8 @@ class ReplicaManager(val config: KafkaConfig,
         return Some(createLogReadResult(e))
     }
 
-    val remoteFetch = new DelayedRemoteFetch(remoteFetchTask, remoteFetchResult, remoteFetchInfo,
+    val timeout = config.requestTimeoutMs.toLong
+    val remoteFetch = new DelayedRemoteFetch(remoteFetchTask, remoteFetchResult, remoteFetchInfo, timeout,
```
Comment (kamalcph) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:1342]: Passing the full request timeout to the `DelayedRemoteFetch`, we can reduce the time spent by the handler from the `requestTimeout` to reach up-to here to expire the delayedRemoteFetch requests immediately if required.

**Comment by divijvaidya** [comment]
Could you please help me understand how this change works with `fetch.max.wait.ms` from a user perspective i.e. what happens when we are retrieving data from both local & remote in a single fetch call?

Also, wouldn't this change user clients? Asking because prior to this change users were expecting a guaranteed response within `fetch.max.wait.ms` = 500ms but now they might not receive a response until 40s `request.timeout.ms`. If the user has configured their application timeouts to according to `fetch.max.wait.ms`, this change will break my application.

**Comment by kamalcph** [comment]
> Could you please help me understand how this change works with fetch.max.wait.ms from a user perspective i.e. what happens when we are retrieving data from both local & remote in a single fetch call?

`fetch.max.wait.ms` timeout is applicable only when there is no enough data (`fetch.min.bytes`) to respond back to the client. This is a special case where we are reading the data from both local and remote, the FETCH request has to wait for the tail latency which is a combined latency of reading from both local and remote storage. 

Note that we always read from only one remote partition up-to `max.partition.fetch.bytes` even-though there is bandwidth available in the FETCH response (`fetch.max.bytes`) and the client rotates the partition order in the next FETCH request so that next partitions are served.

> Also, wouldn't this change user clients? Asking because prior to this change users were expecting a guaranteed response within fetch.max.wait.ms = 500ms but now they might not receive a response until 40s request.timeout.ms. If the user has configured their application timeouts to according to fetch.max.wait.ms, this change will break my application.

`fetch.max.wait.ms` doesn't guarantee a response within this timeout. The client expires the request only when it exceeds the `request.timeout.ms` of 30 seconds (default). The time taken to serve the FETCH request can be higher than the `fetch.max.wait.ms` due to slow hard-disk, sector errors in disk and so on.

The [FetchRequest.json](https://sourcegraph.com/github.com/apache/kafka/-/blob/clients/src/main/resources/common/message/FetchRequest.json) doesn't expose the client configured request timeout, so we are using the default server request timeout of 30 seconds. Otherwise, we can introduce one more config `fetch.remote.max.wait.ms` to define the delay timeout for DelayedRemoteFetch requests. We need to decide whether to keep this config in the client/server since the server operator may need to tune this config for all the clients if the remote storage degrades and latency to serve the remote FETCH requests is high.

**Comment by showuon** [comment]
I understand the problem you're trying to solve, but using the server default request timeout doesn't make sense to me. It will break the contract of fetch protocol that `fetch.max.wait.ms` will not be exceeded if no sufficient data in "local" log. I understand the remote read is some kind of grey area about if "data is existed or not", but we have to admit, some users might feel surprised when their fetch doesn't respond in `fetch.max.wait.ms` time. Ideally, we should introduce another config for this remote read waiting purpose, instead of re-using request timeout.

**Comment by showuon** [comment]
Correction: For this:
> some users might feel surprised when their fetch doesn't respond in fetch.max.wait.ms time.

This is wrong. Even if the remote reading is not completed, yet, the fetch request will still return in `fetch.max.wait.ms`. It's just an empty response.

**Comment by kamalcph** [comment]
@showuon @clolov @jeqo 

The patch is ready for review. PTAL. 

Will open a separate PR to emit the RemoteLogReader `FetchRateAndTimeMs` metric.

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java [hunk 2]
```diff
@@ -198,7 +198,7 @@ public class ConsumerConfig extends AbstractConfig {
      * <code>fetch.max.wait.ms</code>
      */
     public static final String FETCH_MAX_WAIT_MS_CONFIG = "fetch.max.wait.ms";
-    private static final String FETCH_MAX_WAIT_MS_DOC = "The maximum amount of time the server will block before answering the fetch request if there isn't sufficient data to immediately satisfy the requirement given by fetch.min.bytes.";
+    private static final String FETCH_MAX_WAIT_MS_DOC = "The maximum amount of time the server will block before answering the fetch request when it is reading near to the tail of the partition (high-watermark) and there isn't sufficient data to immediately satisfy the requirement given by fetch.min.bytes.";
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java]: Should we mention this config is only used for local log fetch? 
Also, we should mention this in the end:
`To tune the remote fetch maximum time wait, please refer to "remote.fetch.max.wait.ms" config.`. WDYT?

**Review by showuon — COMMENTED** [review]
LGTM! Just a minor comment.

**Comment by showuon** [comment]
@kamalcph , could we move the dynamic config change into another PR? I have some comments to it, but that is separate from the original changes.

**Review by showuon — APPROVED** [review]
LGTM!

**Comment by showuon** [comment]
Failed tests are unrelated.

**Pull request merged**